sc/src/main/java/uk/ac/ebi/atlas/experimentpage/JsonExperimentTSnePlotController.java

```diff
@@ -76,14 +81,24 @@ public String tSnePlotWithMetadata(
             @RequestParam(defaultValue = "") String accessKey) {
         Experiment experiment = experimentTrader.getExperiment(experimentAccession, accessKey);
 
-        return GSON.toJson(
-                ImmutableMap.of(
-                        "series",
-                        modelForHighcharts(
-                                "",
-                                new TreeMap<>(
-                                        tSnePlotService.fetchTSnePlotWithMetadata(
-                                                experiment.getAccession(), perplexity, metadata)))));
+        // need to have this check in case there is no metadata for specified experiment
+        if(getTSnePlotMetadata(experiment.getAccession()).equalsIgnoreCase("[]")) {
```

Comment:
Is returning an empty series a requirement for the widget? In the current implementation, with the [added support for missing metadata](https://github.com/ebi-gene-expression-group/atlas/pull/244), you get the t-SNE points which belong to a "Not available" series. 

If you did want to implement some support for requesting non-existent metadata types, I think making use of an appropriate HTTP error code might be better and more informative for the user.

Reply:
Also, when interacting with `JsonArray`s or `JsonObject`s, try to avoid as much as possible doing any kind of string manipulations/comparisons. Try to use methods provided by the class to check if array is empty (for example, `size()`). Alternatively, pass Java objects arround, such as Lists or Maps, and interact with them instead. It will be much less prone to errors.

Reply:
This check was added as a safety net to display a scatter plot with hard coded "k" value in the tsne widget in case there is no metadata. I'm not sure if that can be achieved with the current implementation. I have not tested it yet. 
I guess showing HTTP error codes to the user might not be a great idea with respect to UX design. There is a link to atlas experiment in the footer of the widget which users can refer to if they require more info.

Regarding string manipulation. I have fixed the logic. Now using JsonArray size function

Reply:
When you are making an API endpoint, you have to think of the "UX" of the API. If I am using a REST API, if I get a 404 Not Found error, I immediately know what that means. It is up to the client (in this case the t-SNE widget, I assume), to handle the error code appropriately. If the client handles it badly and reveals the HTTP code to the end user, then yes, it would be bad UX.

I'm not sure what you mean by a hard coded k value. This endpoint is retrieving t-SNE points grouped by metadata, there shouldn't be any k values involved.

Reply:
The endpoint is actually being consumed by [https://github.com/ebi-gene-expression-group/scxa-experiment-page-tsne-plot-view](scxa-experiment-tsne-plot-view). It uses it to fill `cellClusterData` series. The idea to return empty series from backend was not to break the functionality in `scxa-experiment-page-tsne-plot-view` as it throws [https://github.com/ebi-gene-expression-group/scxa-experiment-page-tsne-plot-view/blob/d79deb4bf2b61e3527af29af26f36c70beac2bcb/src/TSnePlotView.js#L109](errors) if there is a bad response.

Regarding hardcoded 'k-value', in 'tsne-widget' we only show one dropdown with just metadata as options so if there is no metadata in an experiment then we still need to show scatter plot but in that case, we have to provide some `k-value`.

Reply:
Okay, in that case, when would the t-SNE plot ever request non-existent metadata, seeing as it is retrieving the metadata from the backend in the first place?

As for the tsne widget, I assume it is doing a separate request to retrieve data grouped by k if no metadata metadata exists, i.e. if the request to `json/experiments/{experimentAccession}/metadata` returns empty results, then retrieve tsne points grouped by k (by doing a separate request). The endpoints shouldn't be responsible for making decisions about empty data, the client should decide what to do about it.